src/blockchain/compact_filters/p2p.rs

```diff
@@ -0,0 +1,539 @@
+            self.consume_addr(new_addresses);
+        }
+    }
+
+    fn consume_addr(&mut self, addrs: Vec<(u32, Address)>) {
+        for (_, addr) in addrs.iter() {
+            if let Ok(socket_addr) = addr.socket_addr() {
+                self.sender
+                    .send(AddressMessage::NewPending(socket_addr))
+                    .unwrap();
+            }
+        }
+    }
+
+    fn work(&mut self) {
+        loop {
+            let next_address = {
+                let mut address_discovery = self.discovery.lock().unwrap();
+                address_discovery.get_next()
+            };
+
+            match next_address {
+                Some(address) => {
+                    let potential_peer = Peer::connect_with_timeout(
+                        address,
+                        Duration::from_secs(3),
+                        Arc::new(Mempool::default()),
+                        self.network,
+                    );
+
+                    if let Ok(peer) = potential_peer {
+                        peer.send(NetworkMessage::GetAddr).unwrap();
+                        self.try_receive_addr(&peer);
+                        self.try_receive_addr(&peer);
+                        self.sender
+                            .send(AddressMessage::NewValid(
+                                address,
+                                peer.get_version().services,
+                            ))
+                            .unwrap();
+                        peer.close().unwrap();
+                    }
+                }
+                None => break,
+            }
+        }
+    }
+}
+
+/// A dedicated cache structure, with cbf/non_cbf separation
+///
+/// [AddressCache] will interface with file i/o
+/// And can te turned into seeds. Generation of seed will put previously cached
+/// cbf addresses at front of the vec, to boost up cbf node findings
+#[derive(Serialize, Deserialize)]
+struct AddressCache {
+    cbf: Vec<SocketAddr>,
+    non_cbf: Vec<SocketAddr>,
+}
+
+impl AddressCache {
+    fn empty() -> Self {
+        Self {
+            cbf: Vec::new(),
+            non_cbf: Vec::new(),
+        }
+    }
+
+    fn from_file(path: &str) -> Option<Self> {
+        let serialized: Result<String, _> = std::fs::read_to_string(path);
+        let serialized = match serialized {
+            Ok(contents) => contents,
+            Err(_) => return None,
+        };
+
+        let address_cache = serde_json::from_str(&serialized).unwrap();
+        Some(address_cache)
+    }
+
+    fn write_to_file(&self, path: &str) -> Result<(), std::io::Error> {
+        let serialized = serde_json::to_string_pretty(&self).unwrap();
+
+        let mut cache_file = File::create(path).unwrap();
+
+        cache_file.write_all(serialized.as_bytes())?;
+
+        Ok(())
+    }
+
+    fn make_seeds(&self) -> VecDeque<SocketAddr> {
+        self.cbf
+            .iter()
+            .chain(self.non_cbf.iter())
+            .copied()
+            .collect()
```

Comment:
I am not sure I understand. 

What we are doing here is ensuring the CBF addresses are in front of the seed vector. In that way, those are the addresses the worker will try first, even before existing seeds.

Because finding cbf peers are much slower than finding regular peers, the discovery speeds up if we try already known cbf addresses first.

Reply:
I was thinking that if the peers are stored in `Vec`s every call to this method would return the same few peers in the same order, unless they are explicitly removed from the cache.

If the caller uses a logic like "as soon as I reach 5 valid peers I'll stop looking" returning the seeds in the same order could lead to the caller just re-using the same peers over and over again, even across shutdowns.

Using `HashSet`s somewhat mitigates this I think.

Reply:
Ya, once we have `HashSet` the ordering is jumbled up. 

But it's still true that it's the same peers from the previous fetch that were cached in the last run.

I think we cannot do away with that repetition (that's the point of a cache anyway), So the manager checks the connection again, see if the peer is alive, see if it's not in the ban list, check weather cbf/non_cbf, and then pass it on to the caller.

So at the end of all of it, we can be sure that the manager didn't just hand over a dead peer just because it had it in the cache last time.